gen3/bin/healthcheck.sh

```diff
@@ -0,0 +1,118 @@
+#
+# TODO: write info here
+#
+
+source "${GEN3_HOME}/gen3/lib/utils.sh"
+gen3_load "gen3/gen3setup"
+
+help() {
+  gen3 help healthcheck
+}
+
+gen3_healthcheck() {
+  # refer to k8s api docs for pod status info
+  # https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.13/#podstatus-v1-core
+  local allPods=$(g3kubectl get pods -o json | \
+    jq -r '[
+      .items[] | {
+        name: .metadata.generateName, phase: .status.phase, reason: .status.reason,
+        created: .metadata.creationTimestamp, waitingContainers: [ .status.containerStatuses[] | { state: .state, ready:.ready} ]
+      }
+    ]')
+
+  local evictedPods=$(echo $allPods | jq -r '.[] | select(.reason == "Evicted")')
+  local pendingPods=$(echo $allPods | jq -r '[.[] | select(.phase == "Pending")]')
+  local unknownPods=$(echo $allPods | jq -r '.[] | select(.phase == "Unknown")')
+  local crashLoopPods=$(echo $allPods | jq -r '.[] | select( .waitingContainers[].state.waiting.reason == "CrashLoopBackOff")')
+  local terminatingPods='[]'
+  local pendingLongPods='[]'
+
+  # check for pods pending for more than 10 minutes
+  while read -r pod; do
+    local podDate=$(echo $pod | jq -r '.created')
+    local startTime=$(date --date="$podDate" '+%s')
+    local secsPassed=$(( $(date '+%s') - $startTime ))
+    if [[ $secsPassed -gt 600 ]]; then
+      pendingLongPods=$(echo $pendingLongPods | jq -r ". += [$pod]")
+    fi
+  done <<< "$(echo $pendingPods | jq -c '.[]')"
+
+  # check for terminating pods
+  # Unfortunately the only way to get termination duration is by grepping the `describe` output
+  while read -r pod; do
+    local statusLine=$(g3kubectl describe pod $(echo $pod | jq -r '.name') | grep "Status:" -m 1)
+    if [[ "$(echo $statusLine | awk '{ print $2 }')" == "Terminating" ]]; then
+      # check how long it's been terminating
+      local s=$(echo $statusLine | grep -oP "\d+(?=s)")
+      local m=$(echo $statusLine | grep -oP "\d+(?=m)")
+      local h=$(echo $statusLine | grep -oP "\d+(?=h)")
+      if [[ $m -gt 10 || $s -gt 300 || $h -gt 0 ]]; then
+        terminatingPods=$(echo $terminatingPods | jq -r ". += [$pod]")
+      fi
+    fi
+  done <<< "$(echo $allPods | jq -c '.[]')"
+
+  # check status of nodes
+  # https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.13/#nodestatus-v1-core
+  local allNodes=$(g3kubectl get nodes -o json | \
+    jq -r '[
+      .items[] | {
+        name: .metadata.labels."kubernetes.io/hostname", conditions: [ .status.conditions[] | { (.type): .status }] | add
+      }
+    ]')
+  local notReadyNodes=$(echo $allNodes | jq -r '.[] | select(.conditions.Ready != "True")')
+
+  # check internet access
+  local statusCode=$(curl -s -o /dev/null -I -w "%{http_code}" https://www.google.com)
+  local internetAccess=true
```

Comment:
Add a second statusCode check with an explicity proxy set to `cloud-proxy.internal.io` - if you set that in a `()` subshell, then it goes away when the subshell completes - ex:
```
(
  export http_proxy="http://cloud-proxy.internal.io:3128"
  export https_proxy=$http_proxy
  statusCode=...
  exit 0 or exit 1 depending on test success
)
explicitProxyResult=$?
```

Reply:
One more thing - it would be cool if you could determine whether the healthcheck is running on the `admin vm` (user just typing `gen3 healthcheck`) or in a pod (cronjob, whatever) on the cluster - we ideally want to test `curl` internet acccess on the cluster - can maybe `kubectl exec ...` into fence, or trick out `gen3 devterm` somehow, or just run a `healthcheck-job.yaml`, and wait for the job to finish, so the healthcheck runs on the cluster ...